pkg/distributor/distributor.go

```diff
@@ -1195,13 +1195,25 @@ func (d *Distributor) push(ctx context.Context, pushReq *push.Request) (*mimirpb
 	}
 
 	keys := append(seriesKeys, metadataKeys...)
+	initialMetadataIndex := len(seriesKeys)
 
 	// we must not re-use buffers now until all DoBatch goroutines have finished,
 	// so set this flag false and pass cleanup() to DoBatch.
 	cleanupInDefer = false
 
 	err = ring.DoBatch(ctx, ring.WriteNoExtend, subRing, keys, func(ingester ring.InstanceDesc, indexes []int) error {
-		err := d.send(localCtx, ingester, req.Timeseries, req.Metadata, req.Source)
+		timeseries := make([]mimirpb.PreallocTimeseries, 0, len(indexes))
+		var metadata []*mimirpb.MetricMetadata
```

Comment:
why not

```
metadata := make([]*mimirpb.MetricMetadata, 0, len(indexes) - initialMetadataIndex)
```

Reply:
I don't think that's correct optimization, and could lead to panics.

Say series keys is `[231, 342, 532, 984, 278, 1873, 784, 827]`, and `metadataKeys` is `[478, 874, 983]`.
Then `keys` is `[231, 342, 532, 984, 278, 1873, 784, 827, 478, 874, 983]`, and `initialMetadataIndex=8`.

Ingester-1 should get values < 500, so `indexes` would be: `[0, 1, 4, 8]` (index indexes into `keys` that are < 500).

We know that indexes below `initialMetadataIndex` refer to the `keys`, but we don't know how many there are. In this case, `len(indexes) - initialMetadataIndex` would even be negative number.

Given that most write requests push series, and not metadata, we optimize for that case: we preallocate slice for timeseries, but not metadata.